Rubberduck.VBEEditor/SafeComWrappers/SafeComWrapper.cs

```diff
@@ -119,20 +124,25 @@ public void Dispose()
             GC.SuppressFinalize(this);
         }
 
+        private object _disposalLockObject = new object();
         private bool _isDisposed;
         protected virtual void Dispose(bool disposing)
         {
-            if (_isDisposed)
+            lock (_disposalLockObject)
             {
-                return;
+                if (_isDisposed)
+                {
+                    return;
+                }
+                _isDisposed = true;
             }
 
             if (disposing && !HasBeenReleased)
             {
                 Release();
```

Comment:
Hmm. We take a lock on the `_isDisposed` to avoid race conditions, yet we don't do the same for the `_rcwReferenceCount`? The condition here tests `disposing` which isn't affected by the `_isDisposed`, so it seems to me we still have a race condition on the `Release()`?

Reply:
The lock guarantees that only ever one invocation of `Dispose` can proceed beyond the if statement. All others will return because `_isDisposed` is already true.

Reply:
I follow now. That said, I wonder if `_comSafe.TryRemove(this)` should be inside the `Remove()`? Also, how will we deal with exceptions that are risen in the section after the `lock` block? Because if the first call results in an exception, the `_isDisposed` is already set, so second and subsequent calls would fail to release this?

Reply:
I do not see how the thread passing beyond the if statement could throw an exception before it reaches the `Release` method, which itself can really only throw on the COM release.